### PR TITLE
V2 of Difficulty Death Scaler

### PR DESCRIPTION
**Pull request by anhgelus**
Look at the project.

### CODE REVIEW

File: src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java [hunk 1]
```diff
@@ -7,67 +7,191 @@
...
+            playerDifficulty.applyModifiers();
+
+            final var bounty = bountyMap.get(newPlayer.getUuid());
+            if (bounty == null) return;
+            bounty.onDeath();
+        });
+
+        ServerPlayConnectionEvents.JOIN.register((handler, sender, server) -> {
+            final var playerDifficulty = getPlayerDifficultyManager(server, handler.player);
+            if (playerDifficulty.kickIfDiedTooMuch()) return;
```
Comment (leo-210) [inline src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:157]: Is there really no better way to do that ? It doesn't seem very elegant to me.

File: src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java [hunk 2]
```diff
@@ -7,67 +7,191 @@
...
     @Override
     public void onInitialize() {
         LOGGER.info("Difficulty Death Scaler started");
 
-        final var difficultyManager = new DifficultyManager();
-
         final LiteralArgumentBuilder<ServerCommandSource> command = literal("difficultydeathscaler");
-        command.then(literal("get").executes(context -> {
+
+        final LiteralArgumentBuilder<ServerCommandSource> globalCommand = literal("global");
```
[the remark this answers is not in the document]
Comment (leo-210) [reply src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:61]: also maybe add an alias for difficultydeathscaler, it's a long name

**Review by leo-210 — CHANGES_REQUESTED** [review]

File: src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java [hunk 3]
```diff
@@ -7,67 +7,191 @@
...
+
+            final var bounty = bountyMap.get(entity.getUuid());
+            if (bounty == null) return;
+            if (damageSource.getAttacker() instanceof final ServerPlayerEntity player) {
+                bounty.onKill(getPlayerDifficultyManager(player.server, player));
+            }
         });
 
-        ServerPlayConnectionEvents.JOIN.register((handler, sender, server) -> difficultyManager.applyHealthModifierToPlayer(handler.player));
+        ServerPlayerEvents.AFTER_RESPAWN.register((oldPlayer, newPlayer, alive) -> {
```
[the remark this answers is not in the document]
Comment (anhgelus) [reply src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:142]: yes

File: src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java (hunk 1 above)
Comment (anhgelus) [reply src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:157]: no because I do not want to ban the player

Comment (leo-210) [reply src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:157]: can't you ban then un-ban the player instead ?

File: src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (anhgelus) [reply src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:61]: > I don't like the name 'global'
> I would prefer just 'get', or 'difficulty' to get the current global difficulty
> And then you use 'player' for the player difficulty

global is more obvious than player, and it follows the "new" Minecraft command scheme, but I can add an alias for that

[the remark this answers is not in the document]
Comment (leo-210) [reply src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:61]: :ok_hand:

File: src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java (hunk 1 above)
Comment (anhgelus) [reply src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:157]: it requires a lot of code:
- creating a new timertask to unban the player (and saving it to support restart)
- saving older timertask to remove them if the player is banned
- unbanning every player banned if the gamerule is changed

Comment (leo-210) [reply src/main/java/world/anhgelus/architectsland/difficultydeathscaler/DifficultyDeathScaler.java:162]: ok

**Pull request merged**